Iceberg tables support

Hi 

We're hitting an issue with the version 0.7.6, while inserting in Iceberg tables, we get the WARN, followed by the next ERROR, which fails the Spark app:
```
022-04-26 18:19:46,164 WARN LineageHarvester: Write extraction failed for: class org.apache.spark.sql.catalyst.plans.logical.AppendData
scala.MatchError: spark_catalog.gns_test.airports_iceberg_2 (of class org.apache.iceberg.spark.source.SparkTable)
	at za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin.za$co$absa$spline$harvester$plugin$embedded$DataSourceV2Plugin$$extractSourceIdFromTable(DataSourceV2Plugin.scala:134)
	at za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin.za$co$absa$spline$harvester$plugin$embedded$DataSourceV2Plugin$$extractSourceIdFromRelation(DataSourceV2Plugin.scala:128)
	at za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$$anonfun$2.applyOrElse(DataSourceV2Plugin.scala:66)
	at za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$$anonfun$2.applyOrElse(DataSourceV2Plugin.scala:51)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:228)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:224)
	at za.co.absa.spline.harvester.builder.write.PluggableWriteCommandExtractor.asWriteCommand(PluggableWriteCommandExtractor.scala:44)
	at za.co.absa.spline.harvester.LineageHarvester.$anonfun$tryExtractWriteCommand$1(LineageHarvester.scala:147)
	at scala.util.Try$.apply(Try.scala:213)
	at za.co.absa.spline.harvester.LineageHarvester.tryExtractWriteCommand(LineageHarvester.scala:147)
	at za.co.absa.spline.harvester.LineageHarvester.harvest(LineageHarvester.scala:64)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler.onSuccess(QueryExecutionEventHandler.scala:44)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$3(SplineQueryExecutionListener.scala:41)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$3$adapted(SplineQueryExecutionListener.scala:41)
	at scala.Option.foreach(Option.scala:407)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$1(SplineQueryExecutionListener.scala:41)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.withErrorHandling(SplineQueryExecutionListener.scala:52)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:39)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:158)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:128)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:117)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:101)
	at org.apache.spark.sql.util.ExecutionListenerBus.postToAll(QueryExecutionListener.scala:128)
	at org.apache.spark.sql.util.ExecutionListenerBus.onOtherEvent(QueryExecutionListener.scala:140)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent(SparkListenerBus.scala:100)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent$(SparkListenerBus.scala:28)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:117)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:101)
	at org.apache.spark.scheduler.AsyncEventQueue.super$postToAll(AsyncEventQueue.scala:105)
	at org.apache.spark.scheduler.AsyncEventQueue.$anonfun$dispatch$1(AsyncEventQueue.scala:105)
	at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at org.apache.spark.scheduler.AsyncEventQueue.org$apache$spark$scheduler$AsyncEventQueue$$dispatch(AsyncEventQueue.scala:100)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.$anonfun$run$1(AsyncEventQueue.scala:96)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1404)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.run(AsyncEventQueue.scala:96)
which is followed by:2022-04-26 07:40:31,104 ERROR Utils: uncaught error in thread spark-listener-group-shared, stopping SparkContext
java.lang.NoClassDefFoundError: org/apache/hadoop/hive/metastore/api/Index
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.getDeclaredMethod(Class.java:2128)
	at scala.reflect.runtime.JavaMirrors$JavaMirror.$anonfun$methodToJava$1(JavaMirrors.scala:1326)
	at scala.reflect.runtime.TwoWayCaches$TwoWayCache.$anonfun$toJava$1(TwoWayCaches.scala:61)
	at scala.reflect.runtime.TwoWayCaches$TwoWayCache.toJava(TwoWayCaches.scala:57)
	at scala.reflect.runtime.JavaMirrors$JavaMirror.methodToJava(JavaMirrors.scala:1322)
	at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jmeth$lzycompute(JavaMirrors.scala:365)
	at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jmeth(JavaMirrors.scala:365)
	at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaVanillaMethodMirror0.jinvokeraw(JavaMirrors.scala:397)
	at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jinvoke(JavaMirrors.scala:373)
	at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaVanillaMethodMirror.apply(JavaMirrors.scala:389)
	at za.co.absa.commons.reflect.ReflectionUtils$.$anonfun$extractFieldValue$5(ReflectionUtils.scala:143)
	at scala.Option.map(Option.scala:230)
	at za.co.absa.commons.reflect.ReflectionUtils$.reflectClass$1(ReflectionUtils.scala:141)
	at za.co.absa.commons.reflect.ReflectionUtils$.reflectClassHierarchy$1(ReflectionUtils.scala:112)
	at za.co.absa.commons.reflect.ReflectionUtils$.extractFieldValue(ReflectionUtils.scala:169)
	at za.co.absa.commons.reflect.ReflectionUtils$.extractFieldValue(ReflectionUtils.scala:184)
	at za.co.absa.spline.harvester.logging.ObjectStructureDumper$.$anonfun$dump$default$2$1(ObjectStructureDumper.scala:36)
	at za.co.absa.spline.harvester.logging.ObjectStructureDumper$.$anonfun$objectToStringRec$3(ObjectStructureDumper.scala:97)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:198)
	at za.co.absa.spline.harvester.logging.ObjectStructureDumper$.objectToStringRec(ObjectStructureDumper.scala:94)
	at za.co.absa.spline.harvester.logging.ObjectStructureDumper$.dump(ObjectStructureDumper.scala:41)
	at za.co.absa.spline.harvester.logging.ObjectStructureLogging.$anonfun$logObjectStructure$1(ObjectStructureLogging.scala:41)
	at scala.util.Try$.apply(Try.scala:213)
	at za.co.absa.spline.harvester.logging.ObjectStructureLogging.logObjectStructure(ObjectStructureLogging.scala:41)
	at za.co.absa.spline.harvester.logging.ObjectStructureLogging.logWarningObjectStructure(ObjectStructureLogging.scala:31)
	at za.co.absa.spline.harvester.logging.ObjectStructureLogging.logWarningObjectStructure$(ObjectStructureLogging.scala:30)
	at za.co.absa.spline.harvester.LineageHarvester.logWarningObjectStructure(LineageHarvester.scala:46)
	at za.co.absa.spline.harvester.LineageHarvester.tryExtractWriteCommand(LineageHarvester.scala:160)
	at za.co.absa.spline.harvester.LineageHarvester.harvest(LineageHarvester.scala:64)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler.onSuccess(QueryExecutionEventHandler.scala:44)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$3(SplineQueryExecutionListener.scala:41)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$3$adapted(SplineQueryExecutionListener.scala:41)
	at scala.Option.foreach(Option.scala:407)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$1(SplineQueryExecutionListener.scala:41)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.withErrorHandling(SplineQueryExecutionListener.scala:52)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:39)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:153)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:127)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:115)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:99)
	at org.apache.spark.sql.util.ExecutionListenerBus.postToAll(QueryExecutionListener.scala:127)
	at org.apache.spark.sql.util.ExecutionListenerBus.onOtherEvent(QueryExecutionListener.scala:133)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent(SparkListenerBus.scala:82)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent$(SparkListenerBus.scala:28)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:115)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:99)
	at org.apache.spark.scheduler.AsyncEventQueue.super$postToAll(AsyncEventQueue.scala:105)
	at org.apache.spark.scheduler.AsyncEventQueue.$anonfun$dispatch$1(AsyncEventQueue.scala:105)
	at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at org.apache.spark.scheduler.AsyncEventQueue.org$apache$spark$scheduler$AsyncEventQueue$$dispatch(AsyncEventQueue.scala:100)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.$anonfun$run$1(AsyncEventQueue.scala:96)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1319)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.run(AsyncEventQueue.scala:96)
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.hive.metastore.api.Index
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$$anon$1.doLoadClass(IsolatedClientLoader.scala:255)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$$anon$1.loadClass(IsolatedClientLoader.scala:244)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 64 more
```
Do you have any ideas or have you encountered this?

Thanks a lot

--- a/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DataSourceV2Plugin.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DataSourceV2Plugin.scala
@@ -25,6 +25,7 @@ import za.co.absa.spline.harvester.plugin.Plugin.{Params, Precedence, ReadNodeIn
 import za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin._
 import za.co.absa.spline.harvester.plugin.{Plugin, ReadNodeProcessing, WriteNodeProcessing}
 
+import java.net.URI
 import javax.annotation.Priority
 import scala.language.reflectiveCalls
 import scala.util.Try
@@ -138,18 +139,25 @@ class DataSourceV2Plugin
       val name = extractFieldValue[AnyRef](metadata, "name")
       SourceIdentifier(Some("cassandra"), s"cassandra:$keyspace:$name")
 
-    case `_: DeltaTableV2`(dt) => extractSourceIdFromDeltaTableV2(dt)
     case `_: DatabricksDeltaTableV2`(dt) => extractSourceIdFromDeltaTableV2(dt)
 
     case `_: FileTable`(ft) =>
       val format = extractFieldValue[String](ft, "formatName").toLowerCase
       val paths = extractFieldValue[Seq[String]](ft, "paths")
       SourceIdentifier(Some(format), paths: _*)
+
+    case `_: TableV2`(tv2) => extractSourceIdFromDeltaTableV2(tv2)
   }
 
   private def extractSourceIdFromDeltaTableV2(table: AnyRef): SourceIdentifier = {
     val tableProps = extractFieldValue[java.util.Map[String, String]](table, "properties")
-    val uri = tableProps.get("location")
+    val location = tableProps.get("location")
+    val uri =
+      if(URI.create(location).getScheme == null) {
+        s"file:$location"
+      } else {
+        location
+      }
     val provider = tableProps.get("provider")
     SourceIdentifier(Some(provider), uri)
   }
@@ -183,13 +191,14 @@ object DataSourceV2Plugin {
   object `_: FileTable` extends SafeTypeMatchingExtractor[AnyRef](
     "org.apache.spark.sql.execution.datasources.v2.FileTable")
 
-  object `_: DeltaTableV2` extends SafeTypeMatchingExtractor[AnyRef](
-    "org.apache.spark.sql.delta.catalog.DeltaTableV2")
-
   object `_: DatabricksDeltaTableV2` extends SafeTypeMatchingExtractor[AnyRef](
     "com.databricks.sql.transaction.tahoe.catalog.DeltaTableV2")
 
   object `_: CassandraTable` extends SafeTypeMatchingExtractor[AnyRef](
     "com.datastax.spark.connector.datasource.CassandraTable")
+
+  object `_: TableV2` extends SafeTypeMatchingExtractor[AnyRef](
+    "org.apache.spark.sql.connector.catalog.Table")
+
 
 }

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -147,6 +147,28 @@
 
     <profiles>
         <profile>
+            <id>spark-3.0</id>
+            <dependencies>
+                <dependency>
+                    <groupId>org.apache.iceberg</groupId>
+                    <artifactId>iceberg-spark3-runtime</artifactId>
+                    <version>0.13.1</version>
+                    <scope>test</scope>
+                </dependency>
+            </dependencies>
+        </profile>
+        <profile>
+            <id>spark-3.1</id>
+            <dependencies>
+                <dependency>
+                    <groupId>org.apache.iceberg</groupId>
+                    <artifactId>iceberg-spark-runtime-3.1_2.12</artifactId>
+                    <version>0.13.1</version>
+                    <scope>test</scope>
+                </dependency>
+            </dependencies>
+        </profile>
+        <profile>
             <id>spark-3.2</id>
             <dependencies>
                 <dependency>

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -145,6 +145,20 @@
         </dependency>
     </dependencies>
 
+    <profiles>
+        <profile>
+            <id>spark-3.2</id>
+            <dependencies>
+                <dependency>
+                    <groupId>org.apache.iceberg</groupId>
+                    <artifactId>iceberg-spark-runtime-3.2_2.12</artifactId>
+                    <version>0.13.1</version>
+                    <scope>test</scope>
+                </dependency>
+            </dependencies>
+        </profile>
+    </profiles>
+
     <build>
         <plugins>
             <plugin>

--- a/integration-tests/src/test/scala/za/co/absa/spline/IcebergDSV2Spec.scala
+++ b/integration-tests/src/test/scala/za/co/absa/spline/IcebergDSV2Spec.scala
@@ -1,0 +1,96 @@
+/*
+ * Copyright 2020 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package za.co.absa.spline
+
+import org.apache.spark.SPARK_VERSION
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+import za.co.absa.commons.scalatest.ConditionalTestTags.ignoreIf
+import za.co.absa.commons.version.Version.VersionStringInterpolator
+import za.co.absa.spline.test.fixture.spline.SplineFixture
+import za.co.absa.spline.test.fixture.{SparkDatabaseFixture, SparkFixture}
+
+class IcebergDSV2Spec extends AsyncFlatSpec
+  with Matchers
+  with SparkFixture
+  with SplineFixture
+  with SparkDatabaseFixture {
+
+  it should "support Write to iceberg" taggedAs ignoreIf(ver"$SPARK_VERSION" < ver"3.0.0") in
+    withRestartingSparkContext {
+      withCustomSparkSession(_
+        .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions")
+        .config("spark.sql.catalog.iceberg_catalog", "org.apache.iceberg.spark.SparkCatalog")
+        .config("spark.sql.catalog.iceberg_catalog.type", "hadoop")
+        .config("spark.sql.catalog.iceberg_catalog.warehouse", warehouseDir)
+      ) { implicit spark =>
+        withLineageTracking { lineageCaptor =>
+          spark.sql("CREATE TABLE iceberg_catalog.foo (id bigint, data string) USING iceberg;")
+
+          spark
+            .catalog
+            .listDatabases
+            .show(truncate = false)
+
+          for {
+            (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
+              spark.sql("INSERT INTO iceberg_catalog.foo VALUES (1, 'a'), (2, 'b'), (3, 'c');")
+            }
+          } yield {
+            plan1.id.get shouldEqual event1.planId
+            plan1.operations.write.append shouldBe true
+            plan1.operations.write.extra.get("destinationType") shouldBe Some("iceberg")
+            plan1.operations.write.outputSource shouldBe s"file:$warehouseDir/foo"
+          }
+        }
+      }
+    }
+
+  it should "support read from iceberg" taggedAs ignoreIf(ver"$SPARK_VERSION" < ver"3.0.0") in
+    withRestartingSparkContext {
+      withCustomSparkSession(_
+        .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions")
+        .config("spark.sql.catalog.iceberg_catalog", "org.apache.iceberg.spark.SparkCatalog")
+        .config("spark.sql.catalog.iceberg_catalog.type", "hadoop")
+        .config("spark.sql.catalog.iceberg_catalog.warehouse", warehouseDir)
+      ) { implicit spark =>
+        withLineageTracking { lineageCaptor =>
+          spark.sql("CREATE TABLE iceberg_catalog.bar (id bigint, data string) USING iceberg;")
+          spark.sql("CREATE TABLE iceberg_catalog.baz (id bigint, data string) USING iceberg;")
+
+          spark
+            .catalog
+            .listDatabases
+            .show(truncate = false)
+
+          for {
+            (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
+              spark.sql("INSERT INTO iceberg_catalog.bar VALUES (1, 'a'), (2, 'b'), (3, 'c');")
+            }
+            (plan2, Seq(event2)) <- lineageCaptor.lineageOf {
+              spark.sql("INSERT INTO iceberg_catalog.baz SELECT * FROM iceberg_catalog.bar;")
+            }
+          } yield {
+            plan2.id.get shouldEqual event2.planId
+            plan2.operations.reads.get(0).extra.get("sourceType") shouldBe Some("iceberg")
+            plan2.operations.reads.get(0).inputSources(0) shouldBe s"file:$warehouseDir/bar"
+          }
+        }
+      }
+    }
+}

--- a/integration-tests/src/test/scala/za/co/absa/spline/IcebergDSV2Spec.scala
+++ b/integration-tests/src/test/scala/za/co/absa/spline/IcebergDSV2Spec.scala
@@ -42,11 +42,6 @@ class IcebergDSV2Spec extends AsyncFlatSpec
         withLineageTracking { lineageCaptor =>
           spark.sql("CREATE TABLE iceberg_catalog.foo (id bigint, data string) USING iceberg;")
 
-          spark
-            .catalog
-            .listDatabases
-            .show(truncate = false)
-
           for {
             (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
               spark.sql("INSERT INTO iceberg_catalog.foo VALUES (1, 'a'), (2, 'b'), (3, 'c');")
@@ -73,13 +68,8 @@ class IcebergDSV2Spec extends AsyncFlatSpec
           spark.sql("CREATE TABLE iceberg_catalog.bar (id bigint, data string) USING iceberg;")
           spark.sql("CREATE TABLE iceberg_catalog.baz (id bigint, data string) USING iceberg;")
 
-          spark
-            .catalog
-            .listDatabases
-            .show(truncate = false)
-
           for {
-            (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
+            (_, Seq(_)) <- lineageCaptor.lineageOf {
               spark.sql("INSERT INTO iceberg_catalog.bar VALUES (1, 'a'), (2, 'b'), (3, 'c');")
             }
             (plan2, Seq(event2)) <- lineageCaptor.lineageOf {

